pkg/syncer/controllermanager/controllermanager.go

```diff
@@ -0,0 +1,193 @@
+	informerSource     InformerSource
+	managedControllers map[string]ManagedController
+	startedControllers map[string]context.CancelFunc
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *ControllerManager) Start(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), c.name)
+	logger.Info("Starting controller manager")
+	defer logger.Info("Shutting down controller manager")
+
+	go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	<-ctx.Done()
+}
+
+func (c *ControllerManager) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *ControllerManager) processNextWorkItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	c.updateControllers(ctx)
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *ControllerManager) updateControllers(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+	controllersToStart := map[string]CreateControllerFunc{}
+controllerLoop:
+	for controllerName, managedController := range c.managedControllers {
+		requiredGVRs := managedController.RequiredGVRs
+		for _, gvr := range requiredGVRs {
+			if informer, known := c.informerSource.Informer(gvr); !known {
+				continue controllerLoop
+			} else if !informer.HasSynced() {
+				logger.V(2).Info("waiting for the informer to be synced before starting controller", "gvr", gvr, "controller", controllerName)
+				c.queue.AddAfter("resync", time.Second)
+				continue controllerLoop
+			}
+		}
+		controllersToStart[controllerName] = managedController.Create
+	}
+
+	// Create and start missing controllers that have their required GVRs synced
+	newlyStartedControllers := map[string]context.CancelFunc{}
+	for controllerName, create := range controllersToStart {
+		if _, ok := c.startedControllers[controllerName]; ok {
+			// The controller is already started
+			continue
+		}
+
+		// Create the controller
+
+		start, err := create(ctx)
+		if err != nil {
+			logger.Error(err, "failed creating controller", "controller", controllerName)
+		}
+
+		// Start the controller
+		controllerContext, cancelFunc := context.WithCancel(ctx)
+		go start(controllerContext)
+		newlyStartedControllers[controllerName] = cancelFunc
+	}
+
+	// Remove obsolete controllers that don't have their required GVRs anymore
+	for controllerName, cancelFunc := range c.startedControllers {
+		if _, ok := controllersToStart[controllerName]; ok {
+			// The controller is still expected => don't remove it
+			continue
+		}
+		// The controller should not be running
+		// Stop it and remove it from the list of started controllers
+		cancelFunc()
+		delete(c.startedControllers, controllerName)
+	}
+
+	// Add missing controllers that were created and started above
+	for controllerName, cancelFunc := range newlyStartedControllers {
+		c.startedControllers[controllerName] = cancelFunc
```

Comment:
why not set it directly above?

Reply:
right.

Reply:
Fixed in commit da89d306186c852a5f9f9ab2a3a549f896f5f7e0